relentless/ensemble.py

```diff
@@ -28,94 +73,96 @@ def __init__(self, types, T, P=None, V=None, mu={}, N={}, kB=1.0, conjugates=Non
             raise ValueError('Either P or V must be set.')
 
         # mu-N, must be set by type
-        self._mu = core.TypeDict(self.types)
-        self._N = core.TypeDict(self.types)
+        self._mu = core.FixedKeyDict(keys=self.types)
+        self._N = core.FixedKeyDict(keys=self.types)
         for t in self.types:
             if (t not in mu or mu[t] is None) and (t not in N or N[t] is None):
                 raise ValueError('Either mu or N must be set for type {}.'.format(t))
+            if t in N and N[t] is not None and not isinstance(N[t], int):
+                raise TypeError('All values of N must be integers.')
             if t in N:
                 self._N[t] = N[t]
             if t in mu:
                 self._mu[t] = mu[t]
 
-        # conjugates can be specified (and assumed correct), or they can be deduced
-        if conjugates is not None:
-            self._conjugates = tuple(conjugates)
-        else:
-            if self.P is not None and self.V is not None:
-                raise ValueError('Both P and V cannot be set.')
-            for t in self.types:
-                if self.mu[t] is not None and self.N[t] is not None:
-                    raise ValueError('Both mu and N cannot be set for type {}.'.format(t))
-
-            # build the set of conjugate variables from the constructor
-            conjugates = []
-            if self.P is None:
-                conjugates.append('P')
-            if self.V is None:
-                conjugates.append('V')
-            for t in self.types:
-                if self.mu[t] is None:
-                    conjugates.append('mu_{}'.format(t))
-                if self.N[t] is None:
-                    conjugates.append('N_{}'.format(t))
-            self._conjugates = tuple(conjugates)
+        # conjugates are deduced from the specified parameters
+        if self.P is not None and self.V is not None:
+            raise ValueError('Both P and V cannot be set.')
+        for t in self.types:
+            if self.mu[t] is not None and self.N[t] is not None:
+                raise ValueError('Both mu and N cannot be set for type {}.'.format(t))
```

Comment:
This part is already error checking the type keys. So, you could skip this check above and just use this one.